moveit_core/trajectory_processing/test/robot_trajectory_benchmark.cpp

```diff
@@ -0,0 +1,127 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Mario Prats */
+
+// To run this benchmark, 'cd' to the build/moveit_core/trajectory_processing directory and directly run the binary.
+
+#include <benchmark/benchmark.h>
+#include <moveit/robot_model/robot_model.h>
+#include <moveit/robot_state/robot_state.h>
+#include <moveit/robot_trajectory/robot_trajectory.h>
+#include <moveit/utils/robot_model_test_utils.h>
+#include <moveit/trajectory_processing/time_optimal_trajectory_generation.h>
+
+// Robot and planning group to use in the benchmarks.
+constexpr char TEST_ROBOT[] = "panda";
+constexpr char TEST_GROUP[] = "panda_arm";
+
+// Benchmark manual creation of a trajectory with a given number of waypoints.
+// This includes creating and updating the individual RobotState's.
+static void RobotTrajectoryCreate(benchmark::State& st)
+{
+  int n_states = st.range(0);
+  const moveit::core::RobotModelPtr& robot_model = moveit::core::loadTestingRobotModel(TEST_ROBOT);
+
+  // Make sure the group exists, otherwise exit early with an error.
+  if (!robot_model->hasJointModelGroup(TEST_GROUP))
+  {
+    st.SkipWithError("The planning group doesn't exist.");
+    return;
+  }
+  auto* group = robot_model->getJointModelGroup(TEST_GROUP);
+
+  // Robot state.
+  moveit::core::RobotState robot_state(robot_model);
+  robot_state.setToDefaultValues();
+
+  for (auto _ : st)
+  {
+    auto trajectory = std::make_shared<robot_trajectory::RobotTrajectory>(robot_model, group);
+    for (int i = 0; i < n_states; ++i)
+    {
+      moveit::core::RobotState robot_state_waypoint(robot_state);
+      Eigen::VectorXd joint_values = Eigen::VectorXd::Constant(group->getActiveVariableCount(), std::sin(0.01 * i));
```

Comment:
This seemed a bit "magical" to me at the outset -- can this have a bit more constants/explanation that we're just doing a sine sweep of the joint angles?

Reply:
Added a comment and made the `sin` and `duration_from_previous` more visible.